[SR-891] Creating mixed-type Arrays on Linux from literal fails

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-891      |
   |Radar             | None         |
   |Original Reporter | czechboy0 (JIRA User)      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

Swift: DEVELOPMENT-SNAPSHOT-2016-03-01-a  
OS: Ubuntu 14.04


</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, TypeChecker        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 7c3af9159fce9712b3d1400abbab800d

  </details>





**Issue Description:**


I'm aware of the lack of bridging on Linux, however even this works on OS X and doesn't on Linux:

-   open the REPL

-   \`import Foundation\`

-   \`let json: \[Any\] = \[1,\[2,\[3\]\]\]\`  
    This  
    a) succeeds on OS X and returns a valid Swift Array of type \[Any\]  
    b) fails on Linux with

\`\`\`  
repl.swift:2:22: error: contextual type 'protocol\<\>' cannot be used with array literal  
let json: \[Any\] = \[1,\[2,\[3\]\]\]  
^\~\~\~\~\~\~  
\`\`\`

What is the suggested workaround, other than guiding the type inference by changing the array to \`\[1,\[2,\[3\]\] as \[Any\]\]\`? Is this actually a bug or am I doing something unexpected?

--- a/include/swift/Sema/ConstraintSystem.h
+++ b/include/swift/Sema/ConstraintSystem.h
@@ -5646,7 +5646,7 @@ public:
     return false;
   }
 
-  bool isTooComplex(SmallVectorImpl<Solution> const &solutions) {
+  bool isTooComplex(ArrayRef<Solution> solutions) {
     if (isAlreadyTooComplex.first)
       return true;
 

--- a/lib/Sema/CSRanking.cpp
+++ b/lib/Sema/CSRanking.cpp
@@ -1453,6 +1453,10 @@ ConstraintSystem::findBestSolution(SmallVectorImpl<Solution> &viable,
       bestIdx = i;
       break;
     }
+
+    // Give up if we're out of time.
+    if (isTooComplex(/*solutions=*/{}))
+      return None;
   }
 
   // Make sure that our current best is better than all of the solved systems.
@@ -1483,6 +1487,10 @@ ConstraintSystem::findBestSolution(SmallVectorImpl<Solution> &viable,
       ambiguous = true;
       break;
     }
+
+    // Give up if we're out of time.
+    if (isTooComplex(/*solutions=*/{}))
+      return None;
   }
 
   // If the result was not ambiguous, we're done.

--- a/lib/Sema/CSRanking.cpp
+++ b/lib/Sema/CSRanking.cpp
@@ -1405,6 +1405,10 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
 Optional<unsigned>
 ConstraintSystem::findBestSolution(SmallVectorImpl<Solution> &viable,
                                    bool minimize) {
+  // Don't spend time filtering solutions if we already hit a threshold.
+  if (isTooComplex(viable))
+    return None;
+
   if (viable.empty())
     return None;
   if (viable.size() == 1)

--- a/lib/Sema/ConstraintSystem.cpp
+++ b/lib/Sema/ConstraintSystem.cpp
@@ -3828,6 +3828,10 @@ SolutionResult ConstraintSystem::salvage() {
     // Solve the system.
     solveImpl(viable);
 
+    // If we hit a threshold, we're done.
+    if (isTooComplex(viable))
+      return SolutionResult::forTooComplex(getTooComplexRange());
+
     // Before removing any "fixed" solutions, let's check
     // if ambiguity is caused by fixes and diagnose if possible.
     if (diagnoseAmbiguityWithFixes(viable))
@@ -3872,9 +3876,6 @@ SolutionResult ConstraintSystem::salvage() {
 
     // Fall through to produce diagnostics.
   }
-
-  if (isTooComplex(viable))
-    return SolutionResult::forTooComplex(getTooComplexRange());
 
   // Could not produce a specific diagnostic; punt to the client.
   return SolutionResult::forUndiagnosedError();

--- a/test/Constraints/argument_matching.swift
+++ b/test/Constraints/argument_matching.swift
@@ -1789,3 +1789,16 @@ func rdar93922410(_ completion: (Int?) -> Void) { // expected-note {{'completion
     return completion() // expected-error {{missing argument for parameter #1 in call}}
   }
 }
+
+// https://github.com/apple/swift/issues/43509
+do {
+  func myAssertionFailure(
+    _ message: @autoclosure () -> String = String(),
+    file: StaticString = #file, line: UInt = #line
+  ) {}
+
+  let x: Int?
+  myAssertionFailure(x != nil, "error")
+  // expected-error@-1 {{cannot convert value of type 'Bool' to expected argument type 'String'}}
+  // expected-error@-2 {{missing argument label 'file:' in call}}
+}

--- a/test/Constraints/array_literal.swift
+++ b/test/Constraints/array_literal.swift
@@ -145,10 +145,13 @@ func defaultToAny(i: Int, s: String) {
   let _: [Any] = [1, "a", 3.5]
   let _: [Any] = [1, "a", [3.5, 3.7, 3.9]]
   let _: [Any] = [1, "a", [3.5, "b", 3]]
+  let _: [Any] = [1, [2, [3]]]
   
   let _: [Any?] = [1, "a", nil, 3.5]
   let _: [Any?] = [1, "a", nil, [3.5, 3.7, 3.9]]
   let _: [Any?] = [1, "a", nil, [3.5, "b", nil]]
+  let _: [Any?] = [1, [2, [3]]]
+  let _: [Any?] = [1, nil, [2, nil, [3]]]
 
   let a6 = [B(), C()]
   let _: Int = a6 // expected-error{{value of type '[A]'}}

--- a/test/Constraints/generics.swift
+++ b/test/Constraints/generics.swift
@@ -942,3 +942,26 @@ do {
     }
   }
 }
+
+// https://github.com/apple/swift/issues/43527
+do {
+  struct Box<Contents, U> {}
+
+  class Sweets {}
+  class Chocolate {}
+
+  struct Gift<Contents> {
+    // expected-note@-1 2 {{arguments to generic parameter 'Contents' ('Chocolate' and 'Sweets') are expected to be equal}}
+
+    init(_: Box<[Contents], Never>) {}
+  }
+
+  let box = Box<[Chocolate], Never>()
+
+  var g1: Gift<Sweets>
+  g1 = Gift<Chocolate>(box)
+  // expected-error@-1 {{cannot assign value of type 'Gift<Chocolate>' to type 'Gift<Sweets>'}}
+
+  let g2: Gift<Sweets> = Gift<Chocolate>(box)
+  // expected-error@-1 {{cannot assign value of type 'Gift<Chocolate>' to type 'Gift<Sweets>'}}
+}

--- a/test/stdlib/RuntimeObjC.swift
+++ b/test/stdlib/RuntimeObjC.swift
@@ -304,9 +304,18 @@ Runtime.test("isBridgedVerbatimToObjectiveC") {
 
 //===----------------------------------------------------------------------===//
 
+protocol SomeNativeProto {}
+
 class SomeClass {}
 @objc class SomeObjCClass {}
-class SomeNSObjectSubclass : NSObject {}
+
+class SomeNSObjectSubclass : NSObject, SomeNativeProto {}
+extension SomeNativeProto {
+  // https://github.com/apple/swift/issues/43154
+  func expectSelfTypeNameEqual(to typeName: String) {
+    expectEqual(typeName, _typeName(type(of: self)))
+  }
+}
 
 Runtime.test("typeName") {
   expectEqual("a.SomeObjCClass", _typeName(SomeObjCClass.self))
@@ -321,6 +330,8 @@ Runtime.test("typeName") {
 
   a = NSObject()
   expectEqual("NSObject", _typeName(type(of: a)))
+
+  SomeNSObjectSubclass().expectSelfTypeNameEqual(to: "a.SomeNSObjectSubclass")
 }
 
 class GenericClass<T> {}
@@ -571,7 +582,7 @@ Reflection.test("Class/ObjectiveCBase/Default") {
     expectEqual(expected, output)
   }
 }
-protocol SomeNativeProto {}
+
 @objc protocol SomeObjCProto {}
 extension SomeClass: SomeObjCProto {}
 

--- a/validation-test/Sema/type_checker_perf/fast/issue43386_array_of_ranges.swift.gyb
+++ b/validation-test/Sema/type_checker_perf/fast/issue43386_array_of_ranges.swift.gyb
@@ -1,0 +1,12 @@
+// RUN: %scale-test --begin 1 --end 15 --step 1 --select NumLeafScopes %s -Xfrontend=-solver-expression-time-threshold=1 -Xfrontend=-verify
+// REQUIRES: asserts, no_asan
+
+// https://github.com/apple/swift/issues/43386
+
+// expected-error@+1 {{heterogeneous collection literal could only be inferred to '[Any]'; add explicit type annotation if this is intentional}}
+let _ = [
+%for i in range(0, N):
+  0...1,
+  0..<1,
+%end
+]

--- a/validation-test/Sema/type_checker_perf/slow/issue43369.swift
+++ b/validation-test/Sema/type_checker_perf/slow/issue43369.swift
@@ -1,0 +1,23 @@
+// RUN: %target-typecheck-verify-swift -solver-expression-time-threshold=5
+// REQUIRES: tools-release,no_asan
+
+// Hits the default memory threshold
+// https://github.com/apple/swift/issues/43369
+
+struct A {}
+struct B {}
+
+func - (lhs: A, rhs: A) -> B {}
+func / (lhs: B, rhs: Double) -> B {}
+func - (lhs: B, rhs: B) -> B {}
+
+do {
+  let a0 = A()
+  let a1 = A()
+
+  let t0 = 0
+  let t1 = 0
+
+  // expected-error@+1 {{the compiler is unable to type-check this expression in reasonable time}}
+  let _ = (a0 - a0) / (t0 - t0) - (a1 - a1) / (t1 - t1)
+}

--- a/validation-test/Sema/type_checker_perf/slow/nil_coalescing_dictionary_values.swift.gyb
+++ b/validation-test/Sema/type_checker_perf/slow/nil_coalescing_dictionary_values.swift.gyb
@@ -1,0 +1,10 @@
+// RUN: %scale-test --invert-result --begin 1 --end 8 --step 1 --select NumLeafScopes %s -Xfrontend=-typecheck
+// REQUIRES: asserts, no_asan
+
+let x: Int?
+
+let _ = [
+%for i in range(0, N):
+  "k" : x ?? 0,
+%end
+]

--- a/validation-test/compiler_crashers_2_fixed/issue43442.swift
+++ b/validation-test/compiler_crashers_2_fixed/issue43442.swift
@@ -1,0 +1,13 @@
+// RUN: %target-swift-frontend %s -typecheck
+
+// https://github.com/apple/swift/issues/43442
+
+func quickSort<T: Comparable>(_ a: [T]) -> [T] {
+  if a.isEmpty {
+    return a
+  } else {
+    let head = a[0]
+    let body = a[1..<a.count]
+    return quickSort(body.filter({$0 < head})) + [head] + quickSort(body.filter({$0 >= head}))
+  }
+}

